crates/merkle-tree/examples/global_tree_diff.rs

```diff
@@ -0,0 +1,220 @@
+use std::collections::HashMap;
+use std::time::Instant;
+
+use anyhow::Context;
+use bitvec::prelude::Msb0;
+use bitvec::vec::BitVec;
+use clap::Parser;
+use pathfinder_merkle_tree::merkle_tree::MerkleTree;
+use pathfinder_merkle_tree::PedersenHash;
+use stark_hash::Felt;
+
+#[derive(Parser, Debug)]
+#[clap(name = "Global Tree Diff")]
+#[clap(about = "Diffs the leaves in the global state table", long_about = None)]
+struct Cli {
+    #[clap(long)]
+    /// Path to the database file
+    database: std::path::PathBuf,
+    #[clap(long)]
+    /// Verify diffs by rebuilding tree and verify roots
+    verify: bool,
+    #[clap(long)]
+    /// The block number to stop at.
+    stop: Option<u64>,
+}
+
+/// This is a tool to explore the global state tree for a given database as a series of key-value diffs.
+///
+/// Each block's global tree is opened fully to expose all leaves. These are then diff'd against the
+/// previous block and the diff is stored as a json file. Since each tree is explored fully for each block
+/// this can become quite slow.
+///
+/// A smarter implementation is possible - for example, by caching counts for each
+/// node in the previous tree (to prevent re-exploring unchanged areas of the tree). This is kept as a future
+/// improvement.
+///
+/// The leaf and diff count for each block is also reported via stdout.
+fn main() -> anyhow::Result<()> {
+    let cli = dbg!(Cli::parse());
+    let db_filepath = cli.database;
+    let verify = cli.verify;
+
+    let folder = std::path::Path::new(&env!("CARGO_MANIFEST_DIR")).join("global tree diff");
```

Comment:
Those spaces in the dir name are not so Linux friendly

Reply:
I should probably just make it configurable or dump in the current directory 🤷